packages/cli/src/project-references.ts

```diff
@@ -0,0 +1,184 @@
+import typescript from 'typescript';
+
+import { getDefinedArray } from './utils.js';
+
+const { createCompilerHost, getOutputFileNames } = typescript;
+
+/**
+ * A dependency graph where each value has a list of dependencies.
+ *
+ * @template Value - The type of the values in the graph.
+ */
+export type DependencyGraph<Value> = Map<Value, Value[]>;
+
+/**
+ * Create a dependency graph from the resolved project references.
+ *
+ * @param resolvedProjectReferences - The resolved project references of the
+ * package that is being built.
+ * @returns The dependency graph.
+ */
+export function createGraph(
+  resolvedProjectReferences: readonly ResolvedProjectReference[],
+): DependencyGraph<ResolvedProjectReference> {
+  const graph: DependencyGraph<ResolvedProjectReference> = new Map();
+
+  for (const projectReference of resolvedProjectReferences) {
+    graph.set(projectReference, getDefinedArray(projectReference.references));
+  }
+
+  return graph;
+}
+
+/**
+ * The result of a topological sort.
+ *
+ * @template Value - The type of the values in the graph.
+ */
+export type TopologicalSortResult<Value> = {
+  /**
+   * The sorted nodes.
+   */
+  stack: Value[];
+
+  /**
+   * Whether there was a cycle in the graph.
+   */
+  cycle: boolean;
+};
+
+/**
+ * Topologically sort a dependency graph, i.e., sort the nodes in the graph such
+ * that all dependencies of a node come before the node itself.
+ *
+ * @param graph - The dependency graph to sort.
+ * @returns The topologically sorted nodes and whether there was a cycle.
+ */
+export function topologicalSort<Value>(
+  graph: DependencyGraph<Value>,
+): TopologicalSortResult<Value> {
+  const stack: Value[] = [];
+  const visited = new Set<Value>();
+  const recursionStack = new Set<Value>();
+  let cycle = false;
+
+  /**
+   * Visit a node in the graph.
+   *
+   * @param node - The node to visit.
+   */
+  function visit(node: Value) {
+    if (recursionStack.has(node)) {
+      cycle = true;
+      return;
+    }
+
+    if (visited.has(node)) {
+      return;
+    }
+
+    recursionStack.add(node);
+    visited.add(node);
+
+    const neighbours = graph.get(node);
+    assert(neighbours !== undefined);
+
+    neighbours.forEach(visit);
+    recursionStack.delete(node);
+    stack.push(node);
+  }
+
+  for (const node of graph.keys()) {
+    visit(node);
+  }
+
+  return {
+    stack,
+    cycle,
+  };
+}
+
+/**
+ * Get the resolved project references from a TypeScript program.
+ *
+ * @param resolvedProjectReferences - The resolved project references of the
+ * package that is being built.
+ * @returns The resolved project references.
+ */
+export function getResolvedProjectReferences(
+  resolvedProjectReferences: ResolvedProjectReference[],
+) {
+  const graph = createGraph(resolvedProjectReferences);
+  const { stack, cycle } = topologicalSort(graph);
+
+  if (cycle) {
+    throw new Error(
+      'Unable to build project references due to a dependency cycle.',
+    );
+  }
```

Comment:
Should we also identify the name of the cyclically dependent package?

Reply:
Maybe just display the cycle itself?

Reply:
Done:

```
Unable to build project references due to a dependency cycle:
- packages/project-1/tsconfig.circular.json -> packages/project-3/tsconfig.circular.json -> packages/project-2/tsconfig.circular.json -> packages/project-1/tsconfig.circular.json
```